### DISCUSSION TITLE
How to bind window global object?

### DISCUSSION THREAD

**Discussion by nikolajevs86** [wasm-bindgen/wasm-bindgen#3238] [Q&A] [answered]
Hi, sorry this maybe a stupid question, but I am stuck for several weeks already.
I have a global object with name SDK, which I can see that exists, but I cannot cast it.
I need to call some methods of this SDK object, so I am trying to bind it by:

```
#[wasm_bindgen]
extern "C" {
#[wasm_bindgen()]
    pub type SDK;

    #[wasm_bindgen(method)]
    pub fn start(this: &SDK);
}

pub fn execute() {
    let sdk_js_value = JsValue::from_str("SDK");
    
    if js_sys::Reflect::has(&js_sys::global(), &sdk_js_value).unwrap() {
        log!("SDK exists");        

        let sdk = js_sys::global().dyn_into::<SDK>().ok().unwrap(); // Here error is thrown
        sdk.start();
    } else {
        log!("SDK does not exists");
    }
}
```

It fail to cast to SDK object and I am not sure why it cannot do it. Can someone suggest something please?

**Comment by Liamolucko** [comment, marked as answer]
It's possible to get it to work the way you're trying to do it, getting the `SDK` from `js_sys::global` and then casting it, but I'd recommend doing it this way instead.

You can import methods from `SDK` by adding `js_namespace = SDK` to a regular function import:

```rust
use wasm_bindgen::prelude::*;

#[wasm_bindgen]
extern "C" {
    #[wasm_bindgen(js_namespace = SDK)]
    fn start();
}

pub fn execute() {
    if js_sys::Reflect::has(&js_sys::global(), &JsValue::from_str("SDK")).unwrap() {
        log!("SDK exists");

        start();
    } else {
        log!("SDK does not exists");
    }
}
```

Calling `start` will run `SDK.start()` on the JS side.

**Reply by nikolajevs86** [reply]
It works, it appeared so easy, thanks a lot!
Just to sum up, so by this approach we are mapping js function of global SDK object, but if we want to use this SDK object properties then we need to cast it to the object using dyn_into, right?

**Reply by Liamolucko** [reply]
Yes, you would need to get a reference to the actual SDK object to get at its properties. Retrieving it from `js_sys::global()` and using `dyn_into` is one way to do that, but you can also get at the object by using a `static`:

```rust
use wasm_bindgen::prelude::*;

#[wasm_bindgen]
extern "C" {
    type Sdk;

    #[wasm_bindgen(method, getter)]
    fn property(this: &Sdk) -> String;

    static SDK: Sdk;
}

pub fn execute() {
    let x = SDK.property();
}
```
